File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/OngoingCallHelper.java
package dev.alenajam.opendialer.feature.inCall.service;

import android.content.Context;
import android.telecom.Call;
import android.telecom.DisconnectCause;
import android.widget.Toast;

import androidx.annotation.Nullable;

public abstract class OngoingCallHelper {
    public static boolean handleDisconnectCause(Context context, Call call) {
        boolean hasCause = false;
        if (isDisconnectedByError(call)) {
            String cause = getDisconnectCauseDesc(call);
            hasCause = cause != null && !cause.isEmpty();
            if (hasCause) {
                Toast.makeText(context, cause, Toast.LENGTH_LONG).show();
            }
        }
        return hasCause;
    }

    @Nullable
    public static String getDisconnectCauseDesc(Call call) {
        CharSequence desc = call.getDetails().getDisconnectCause().getDescription();
        if (desc == null) return null;
        return desc.toString();
    }

    public static boolean isDisconnectedByError(Call call) {
        int code = call.getDetails().getDisconnectCause().getCode();
        return code != DisconnectCause.LOCAL && code != DisconnectCause.REMOTE;
    }
}

File: feature/inCall/src/main/java/dev/alenajam/opendialer/feature/inCall/service/OngoingCallHelper.kt
package dev.alenajam.opendialer.feature.inCall.service

import android.content.Context
import android.os.Build
import android.telecom.Call
import android.telecom.DisconnectCause
import android.widget.Toast

val Call.safeState: Int
    get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        details.state
    } else {
        @Suppress("DEPRECATION")
        state
    }

object OngoingCallHelper {
    @JvmStatic
    fun handleDisconnectCause(context: Context, call: Call): Boolean {
        if (!isDisconnectedByError(call)) return false

        val cause = getDisconnectCauseDesc(call)
        val hasCause = !cause.isNullOrEmpty()
        if (hasCause) {
            Toast.makeText(context, cause, Toast.LENGTH_LONG).show()
        }
        return hasCause
    }

    @JvmStatic
    fun getDisconnectCauseDesc(call: Call): String? =
        call.details.disconnectCause.description?.toString()

    @JvmStatic
    fun isDisconnectedByError(call: Call): Boolean {
        val code = call.details.disconnectCause.code
        return code != DisconnectCause.LOCAL && code != DisconnectCause.REMOTE
    }
}
